shell/server/drivers/mail.js

```diff
@@ -223,15 +240,23 @@ hackSendEmail = (session, email) => {
       };
     }
 
-    if (email.from.address !== grainAddress && email.from.address !== userAddress.address) {
+    if (removeValidPlusSuffix(email.from.address) !== grainAddress && email.from.address !== userAddress.address) {
       throw new Error(
-        "FROM header in outgoing emails need to equal either " + grainAddress + " or " +
+        "FROM header in outgoing emails need to equal either " + grainAddress + " (with optional suffix) or " +
         userAddress.address + ". Yours was: " + email.from.address);
     }
 
     // Unpack fields
     const { from, to, cc, bcc, replyTo, subject, text, html } = email;
 
+    // We allow envelope address to have a suffix as well. In this way apps can use unique envelope
+    // addresses when sending out e-mails which allow better handling of e-mail bounces - app can
+    // match bounce with the sent e-mail.
+    let envelopeFrom = email.envelopeFrom || from.address;
```

Comment:
Currently, `envelopeFrom` will never be present since it is not defined in `email.capnp`. There's no way for an app to fill this in.

We could add it, but I'm not sure if this is the right interface. If the app can't actually set an arbitrary `envelopeFrom` then it is confusing to let them specify one and then fail when they specify the wrong one. Maybe instead we should have a field `bounceId` which is strictly the ID to place after the `+`? Another idea would be to extend the email interfaces such that Sandstorm can capture bounces and make some RPC callback to the app, which might be a nicer interface for apps, but requires a lot more work.

Are you actually anticipating the need to handle bounces in an app? If not then maybe we can punt this for now.